File: 0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.py
# NEETCODE
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        ops = {
            "+": lambda x, y: x + y,
            "-": lambda x, y: x - y,
            "*": lambda x, y: x * y
        }
        
        def backtrack(left, right):
            res = []
            for i in range(left, right + 1):
                op = expression[i]
                if op in ops:
                    num1 = backtrack(left, i - 1)
                    num2 = backtrack(i + 1, right)
                    for n1 in num1:
                        for n2 in num2:
                            res.append(ops[op](n1, n2))
            if res == []:
                res.append(int(expression[left:right + 1]))
            return res
        
        return backtrack(0, len(expression) - 1)

File: 0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {
        unordered_map<char, function<int(int, int)>> ops = {
            {'+', [](int x, int y) {
                return x + y;
            }},
            {'-', [](int x, int y) {
                return x - y;
            }},
            {'*', [](int x, int y) {
                return x * y;
            }}
        };
        
        function<vector<int>(int, int)> backtrack = [&](int l, int r) {
            vector<int> res;
            for (int i = l; i <= r; i++) {
                char op = expression[i];
                if (ops.count(op)) {
                    vector<int> num1 = backtrack(l, i - 1);
                    vector<int> num2 = backtrack(i + 1, r);
                    for (int n1 : num1) {
                        for (int n2 : num2) {
                           res.push_back(ops[op](n1, n2)); 
                        }
                    }
                }
            }
            if (res.empty()) {
                res.push_back(stoi(expression.substr(l, r - l + 1)));
            }
            return res;
        };
        return backtrack(0, expression.length() - 1);
    }
};